azure/Guidelines.md

```diff
@@ -94,74 +94,47 @@ PUT https://blobstore.azure.com/foo.com/acct1/c1/b2?api-version=2014-12-07
+Azure services **MUST** update the version number of their API whenever there is a change to the API, no matter how small.  Customers will "lock the API version" so that their code does not fail when the service introduces new features.  They rely on the fact that an API version is a contract with the services that will never change.
 
-At a high level, changes to the contract of an API constitute a breaking change. Changes that impact backwards compatibility of an API is also considered a breaking change. Anything that would violate the Principle of Least Astonishment is considered a breaking change in Azure. Below are some concrete examples of what constitutes a breaking change. In the below breaking change scenarios, the API version must be changed.
+There are three reasons why a service may issue a breaking change:
 
-#### Existing property is removed
+1. To remove a security vulnerability.
+2. To comply with regulatory requirements.
+3. To deprecate a feature of the service.
 
-If a property called `foo` that was present in v1 of the API needs to be removed, it must be done in a newer API version.
+In each case, prior approval of the Azure REST API review board is required.  In the case of deprecation, follow the API deprecation policy (below).  If the service is using SemVer for versioning, breaking changes constitute a major version change.
 
-#### New property added to response
+#### Examples of breaking changes
 
-If a new property/field is added to the response of an API, the GET-PUT pipeline will be broken. Consider the case where a customer updates the value of a new property "A" from the Azure portal. Another customer does a GET of this resource using the SDK. The SDK will ignore the property since it does not understand it. From the SDK, the customer does a PUT using the model that was returned from the GET. This will overwrite the change made by the first customer from the portal.
+At a high level, any change to the contract of an API constitutes a breaking change. The following is a non-exhaustive list of breaking changes:
 
-#### New required property added to request
+* An existing property is removed.
+* A new property is added to an existing response.
+* A new required property is added to an existing request.
+* A property name is changed.
+* A property type is changed.
+* The default value of a property is changed.
+* The allowed values for an enum is changed.
```

Comment:
If they are only expanded, is it still a breaking change? Perhaps refine by saying "renamed and/or removed".

Reply:
If previously, you had a finite (small) set of possibilities, this may be encoded as an enum in an SDK.  Adding a new possibility would break the decoding of the response in this case.   So, yes - it is a breaking change.  It may be one we are willing to live with, though - that's a good discussion to have.

Reply:
In cognitive services we often improve our capabilities without changing the API. For example, a new language detection model can detect a new language. While we have optional model-version query parameters to make sure the behavior can be controlled by the customers, upgrading an API version would really hinder the ability to evolve. 

Another example, would be the set of error codes returned. It's virtually impossible to try and predict all the possible things that can go wrong with a service on the day you release the API.

Reply:
Is this a reason not to model a property as an enum?  I think it would be useful to have a distinction between 'known good values' and 'an exhaustive enumeration of possibilities' - where the latter is an enum.

Reply:
In Graph we have have the explicit notion of an extensible enum, where customers have to opt in with a header if they are capable of dealing with a changing set of values.  If they don't opt in, then they get a marker value called 'unkownfuturevalue' returned instead of the new value.  So simpler customers get an easy experience with a fixed set plus a marker.  More sophisticated customers get access to the full set of values.   ALl customers do have to deal with unkownfuturevalue, but that is part oft he enum set from day one.

Reply:
To resolve this, I've added an explicit mention of the extensible enum.